Add Package: bigdft-suite

william-dawson opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/bigdft```) and found some lint.

Here's what I've got...


For **recipes/bigdft**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about', 'extra'].
 * license_file entry is missing, but is required.

For **recipes/bigdft**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

Documentation on acceptable licenses can be found [here]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#spdx-identifiers-and-expressions ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/bigdft```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/bigdft```) and found some lint.

Here's what I've got...


For **recipes/bigdft**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [43]

william-dawson commented:
@jan-janssen I see you have made various attempts at building a BigDFT package with conda. I think that this PR should take care of things. If you have any suggestions from your experience, let me know. We still need to make a new release (1.9.4) to finalize the package. Also, if you have some use of BigDFT we're always happy to collaborate.

jan-janssen commented on recipes/bigdft/build.sh:
```diff
@@ -0,0 +1,18 @@
+cp $RECIPE_DIR/patches/* .
+
+# Patches
+patch futile/flib/utils.c utils.patch
+patch bigdft/src/output.f90 output.patch
```
You can patch the sources directly in the `meta.yaml` file.

william-dawson replied:
Thanks, I've made this modification!

jan-janssen commented:
> @jan-janssen I see you have made various attempts at building a BigDFT package with conda. I think that this PR should take care of things. If you have any suggestions from your experience, let me know. We still need to make a new release (1.9.4) to finalize the package. Also, if you have some use of BigDFT we're always happy to collaborate.

Thanks a lot. @chrisjsewell maybe you want to have a look as well

chrisjsewell commented:
This is great cheers @william-dawson 🎉 
It is the last outstanding code, that we were looking to use via conda (https://github.com/marvel-nccr/quantum-mobile/issues/161)

I can have a quick look, but @jan-janssen has a lot more experience of these recipes than me, so mainly I will leave it in he's capable hands to review 😄

Just an FYI, this is the [ansible](https://github.com/ansible/ansible) role I have currently been using to build it with: https://github.com/marvel-nccr/ansible-role-bigdft/blob/master/tasks/main.yml

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/bigdft```) and found some lint.

Here's what I've got...


For **recipes/bigdft**:

 * ``requirements: host: libxc=4.3.4`` must contain a space between the name and the pin, i.e. ``libxc =4.3.4``
 * ``requirements: run: libxc=4.3.4`` must contain a space between the name and the pin, i.e. ``libxc =4.3.4``

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/bigdft```) and found it was in an excellent condition.

luigigenovese commented:
Hello,
thanks for the efforts and replies! I would be very happy to be part of the maintainers for this package (although I did not participate since now to the package recipe creation). Let me know if this is possible. Many thanks! Luigi Genovese

william-dawson commented:
@chrisjsewell That is very interesting. With BigDFT, you will have the same problem as abinit in that it depends on libxc version 4. I suppose this is why the spack people work so hard to mix and match library versions.

chrisjsewell commented:
> That is very interesting.

If you are particularly interested lol, then this is essentially what I was working towards lol, to have all the codes running in a "common" way: https://github.com/marvel-nccr/ansible-role-aiida-cws (and also big thanks to @jan-janssen for getting all the others on conda 🙏 )

william-dawson commented:
@jan-janssen I believe all the tasks are completed. Using `channel_targets` in `conda_build_config.yaml`, I've designated this as a release candidate. If everything looks good to you, I'd like to merge this in, and then start testing out workflows based on installing bigdft from conda. If everything is working well, I'll make a new merge request sometime in September with the official 1.9.3 release of BigDFT.

jan-janssen commented:
@conda-forge/staged-recipes, ready for review

chrisburr commented on recipes/bigdft/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+  version: {{ version }}
+
+source:
+  url: https://gitlab.com/l_sim/bigdft-suite/-/archive/Prerelease1.9.3/bigdft-suite-Prerelease1.9.3.tar.gz
```
The URL should depend on `version` to avoid issues with accidentally deploying a version with the wrong sources.

Also is this a pre-release?

william-dawson replied:
> The URL should depend on version to avoid issues with accidentally deploying a version with the wrong sources.

Done. Thanks for the correction.

> Also is this a pre-release?

That is my plan. We will release version 1.9.3 of BigDFT in about a month and a half, and I'd like to have a pre-release available so all developers can try out the conda approach. I believe that my `conda_build_config.yaml` should properly label this as a pre-release, but let me know if there are fixes I should make.

jaimergp commented on recipes/bigdft/build.sh:
```diff
@@ -0,0 +1,13 @@
+python ../Installer.py -y autogen
+python ../Installer.py -y build -f ../conda.rc
+
+# Activate script
```
Ideally, this should be paired with a `deactivate.sh` script that undoes the effect of the activation script, so the env vars set do not permeate into the user session once the env is deactivated.

william-dawson replied:
Thanks. Do you know of any projects that have well written `deactivate.sh` scripts that unset environment variables? Because it seems like I'd first need to store all of the original environment variable values somewhere (noting that multiple sessions with different initial conditions could be inside the environment at the same time) for restoring. I can write a script to do all this but I'd like to avoid any pitfalls.

jan-janssen replied:
You can search Github for projects in the conda-forge organization which use a `deactivate.sh` script https://github.com/search?q=org%3Aconda-forge+deactivate.sh&type=code

jaimergp replied:
[This is a good example](https://github.com/conda-forge/intel-compiler-repack-feedstock/tree/main/recipe). The idea is to backup potential values that were set for that env var _before_ you activate. Then on deactivation, restore to the backup _or_, if empty, unset.

You will need to take notes of what the `source`d script is doing, though :)

william-dawson replied:
Thank you for the suggestions! Storing in environment variables with the `conda` suffix seems like a good strategy. I've implemented a deactivate script based on this.

jan-janssen commented:
@chrisburr Can you take another look?

carterbox commented on recipes/bigdft/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+    - patches/output.patch
+    - patches/utils.patch
+
+build:
+  number: {{ build }}
+  skip: true  # [win or osx]
```
```suggestion
build:
  number: {{ build }}
  skip: true  # [win or osx]
  # add build string so packages can depend on
  # mpi or nompi variants explicitly:
  # `pkg * mpi_mpich_*` for mpich
  # `pkg * mpi_*` for any mpi
  # `pkg * nompi_*` for no mpi
  {% if mpi != 'nompi' %}
  {% set mpi_prefix = "mpi_" + mpi %}
  {% else %}
  {% set mpi_prefix = "nompi" %}
  {% endif %}
  string: "{{ mpi_prefix }}_py{{ py }}h{{ PKG_HASH }}_{{ build }}"
```
Please indicate the MPI variant in the build string

https://conda-forge.org/docs/maintainer/knowledge_base.html#preferring-a-provider-usually-nompi

william-dawson replied:
There is no `nompi` version of our package, it must be installed with mpi. So in our case this seems unnecessary. Do you recommend instead just putting `{% set mpi_prefix = "mpi_" + mpi %}`? Or to just leave it all out?

carterbox replied:
This is just a suggestion from the docs page. Please modify to fit your use case. You have separate variants for `mpich` and `openmpi` and they should be noted in the build string.

william-dawson replied:
Thanks. I've made the modification accordingly, and will continue to work on the other suggestions.

carterbox commented on recipes/bigdft/test.py:
```diff
@@ -0,0 +1,19 @@
+
+log = code.run(sys=sys, input=inp)
+print(log.energy)
+
```
```suggestion

import CheSS
import futile
import gi
```
Test imports of all your packages.

william-dawson replied:
CheSS is depreciated (I'm put in a pull request to remove it from our main repository) and gi is too experimental to test, but testing the imports explicitly of futile and BigDFT sounds like a good idea to me (better than the error coming from deep in a run). I've made a modification based on this suggestion.

carterbox replied:
The tests in this conda recipe are for *packaging* validation, so their purpose is to check that the artifacts were installed correctly. They are not necessarily a comment on whether a feature of the package is deprecated or considered experimental. If you expect that an end user of this package would still be importing the CheSS module, then you should test it.

carterbox commented on recipes/bigdft/run_test.sh:
```diff
@@ -0,0 +1,2 @@
+source bigdftvars.sh
+python test.py
```
Maybe a few of the many CLI binaries being shipped? Do any of them accept `--help` or `--version` and return?

william-dawson replied:
Thanks. I've added some tests for the CLI apps I know are more regularly used with a `-h` option. I might talk to the other developers about whether all of the helper binaries should be built by default in the future.

carterbox commented on recipes/bigdft/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+    - patches/output.patch
+    - patches/utils.patch
+
+build:
+  number: {{ build }}
```
```suggestion
build:
  run_exports:
     -  {{ pin_subpackage(name|lower) }}
  number: {{ build }}
```
Shipping header files means that you should pin this package according to ABI compatability guarantees.

carterbox commented on recipes/bigdft/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+  home: https://bigdft.org
+  license: GPL-3.0-only
+  license_family: GPL
+  license_file: LICENSE
```
The files in licenses are not being included, but I think globbing works? otherwise, you need to be explicit.

```suggestion
  license_file:
    - LICENSE
    - licenses/*
```

carterbox reviewed (changes requested):
Is `libfmalloc` part of this package or are you redistributing?

Shipping static libraries is not allowed. The following libs are being shipped statically: abinit, atlab, bigdft, CheSS, dics, futile, Galn, orbs, Psolver.

_jhbuild/ is in included in the artifacts, is the supposed to be shipped?

william-dawson commented:
> Is `libfmalloc` part of this package or are you redistributing?
> 
> Shipping static libraries is not allowed. The following libs are being shipped statically: abinit, atlab, bigdft, CheSS, dics, futile, Galn, orbs, Psolver.
> 
> _jhbuild/ is in included in the artifacts, is the supposed to be shipped?

I've set my install script to remove the `_jhbuild` directory as it seems to be only used at build time. I've also modified the configuration lines to build dynamic versions of the libraries (and remove the static versions). This indeed includes `libfmalloc`, which is our Fortran module for managing dynamic memory. 

I think I have addressed all your suggestions, but let me know if I missed or misunderstood something. We appreciate the feedback!

carterbox commented on recipes/bigdft/run_test.sh:
```diff
@@ -0,0 +1,10 @@
+# Setup the environment
```
```suggestion
set -ex

# Check for expected libraries ABI
test -f ${PREFIX}/lib/bigdft/libvol.so
test -f ${PREFIX}/lib/libatlab-1.so.0
test -f ${PREFIX}/lib/libbigdft-1.so.9
test -f ${PREFIX}/lib/libCheSS-1.so.2
test -f ${PREFIX}/lib/libfmalloc-1.so.9
test -f ${PREFIX}/lib/libfutile-1.so.9
test -f ${PREFIX}/lib/liborbs.so.0
test -f ${PREFIX}/lib/libPSolver-1.so.9

# Check for header
test -f ${PREFIX}/include/bigdft.h
test -f ${PREFIX}/include/futile/tree.h
test -f ${PREFIX}/include/atlab/box.h

# Setup the environment
```
It's good practice to do existence checks for libraries with major version. This will cause tests to fail and prevent from breaking downstream packages if the SONAME of these libraries changed unexpectedly.

william-dawson commented on recipes/bigdft/meta.yaml:
```diff
@@ -0,0 +1,82 @@
+
+build:
+  run_exports:
+     -  {{ pin_subpackage(name|lower) }}
```
I see. I didn't understand that point at the time. I think it would be difficult to guarantee that with the current release cycle, so I've updated it to include the minor version as well.

william-dawson commented on recipes/bigdft/test.py:
```diff
@@ -0,0 +1,24 @@
+# Test that we can import everything
+import futile
```
I asked about this upstream, and they said they will fix it as you suggest in a future release.

william-dawson commented:
@carterbox I've gone through the most recent suggestions. For the testing of the dynamic libraries you suggested, I had to remove `liborbs` from the list, as there is an issue with the symbolic link. I will report that upstream. For the CheSS and gi modules, thank you for the clarification. There is essentially a 0% chance someone would include it in their code, so I think it is safe to leave that test out.

william-dawson reviewed (commented):
(see individual comments)

carterbox commented:
@conda-forge/help-python-c, please review

carterbox reviewed (approved):
Apparently, I'm the only reviewer of compiled packages right now. 😆

william-dawson commented:
> Apparently, I'm the only reviewer of compiled packages right now. 😆

Be sure to add that you to CV 😄
